core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java

```diff
@@ -65,9 +68,167 @@ public CachedConnection reserveAny() {
+        try {
+          connection.transport.close();
+        } catch (Exception e) {
+          log.debug("Error closing transport during shutdown", e);
+        }
+      });
+    }
+
+    CachedConnection removeReserved(CachedTTransport transport) {
+      return reserved.remove(transport);
+    }
+  }
+
+  private static class ConnectionPool {
+    final ConcurrentHashMap<ThriftTransportKey,ReentrantLock> locks = new ConcurrentHashMap<>();
+    final ConcurrentHashMap<ThriftTransportKey,CachedConnections> connections = new ConcurrentHashMap<>();
+
+    Set<ThriftTransportKey> getThriftTransportKeys() {
+      return connections.keySet();
+    }
+
+    CachedConnection reserveAny(final ThriftTransportKey key) {
+      return executeWithinLock(key, () -> getOrCreateCachedConnections(key).reserveAny());
+    }
+
+    CachedConnection reserveAnyIfPresent(final ThriftTransportKey key) {
+      return executeWithinLock(key, () -> connections.containsKey(key)
+          ? connections.get(key).reserveAny()
+          : null);
+    }
+
+    void putReserved(final ThriftTransportKey key, final CachedConnection connection) {
+      executeWithinLock(key, () -> getOrCreateCachedConnections(key).reserved.put(
+          connection.transport, connection));
+    }
+
+    boolean returnTransport(final CachedTTransport transport, final List<CachedConnection> toBeClosed) {
+      return executeWithinLock(transport.getCacheKey(), () -> unreserveConnection(transport, toBeClosed));
+    }
+
+    void shutdown() {
+      for (Entry<ThriftTransportKey,CachedConnections> entry : connections.entrySet()) {
+        Lock lock = locks.get(entry.getKey());
+        // Explicitly do not release the lock afterwards in order to prevent other threads from
+        // obtaining it during shutdown. This pool instance is expected to be nullified after this
+        // method is called.
+        lock.lock();
+
+        // Close the transports.
+        entry.getValue().closeAllTransports();
+      }
+    }
+
+    <T> T executeWithinLock(final ThriftTransportKey key, Supplier<T> function) {
+      Lock lock = getLock(key);
+      try {
+        return function.get();
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    void executeWithinLock(final ThriftTransportKey key, Consumer<ThriftTransportKey> consumer) {
+      Lock lock = getLock(key);
+      try {
+        consumer.accept(key);
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    ReentrantLock getLock(final ThriftTransportKey key){
+      return locks.computeIfAbsent(key, k -> new ReentrantLock());
+    }
+
+    CachedConnections getCachedConnections(final ThriftTransportKey key) {
+      return connections.get(key);
+    }
+
+    CachedConnections getOrCreateCachedConnections(final ThriftTransportKey key) {
+      return connections.computeIfAbsent(key, k -> new CachedConnections());
+    }
+
+    boolean unreserveConnection(final CachedTTransport transport,
```

Comment:
When reviewing the code, I found myself wondering if some of these functions were called with a lock?  It made me wonder if these functions that are only called using `executeWithinLock` should be inlined.  However inlining may look awful.  I am not sure about this, just sharing my experience.

Reply:
My primary intent was to ensure that any action that needed to execute within a lock would automatically do so. The lack of inlining was a side benefit. However, if you feel it would improve the overall clarity to inline instead, I can switch it. Alternatively, I can add some documentation for the methods in question to explicitly state that their operations take place within a lock.

Reply:
> Alternatively, I can add some documentation for the methods in question to explicitly state that their operations take place within a lock.

I like that better than inlining.